client/asset/btc/btc.go

```diff
@@ -2361,6 +2411,234 @@ func (btc *baseWallet) fund(keep uint64, minConfs uint32, lockUnspents bool,
+// will be increased by the percentage defined in splitBuffer. If an
+// order requires .1 BTC and the splitBuffer is 5, the order's output will
+// be .105 BTC.
+func (btc *baseWallet) fundMultiWithSplit(keep, maxLock uint64, values []*asset.MultiOrderValue,
+	splitTxFeeRate, maxFeeRate, splitBuffer uint64) ([]asset.Coins, [][]dex.Bytes, uint64, error) {
+	var swapInputSize uint64
+	if btc.segwit {
+		swapInputSize = dexbtc.RedeemP2WPKHInputTotalSize
+	} else {
+		swapInputSize = dexbtc.RedeemP2PKHInputSize
+	}
+
+	var totalOutputRequired uint64
+	requiredForOrders := make([]uint64, len(values))
+	for i, value := range values {
+		req := calc.RequiredOrderFundsAlt(value.Value, swapInputSize, value.MaxSwapCount, btc.initTxSizeBase, btc.initTxSize, maxFeeRate)
+		req = req * (100 + splitBuffer) / 100
+		requiredForOrders[i] = req
+		totalOutputRequired += req
+	}
+
+	var splitTxSizeWithoutInputs uint64 = dexbtc.MinimumTxOverhead
+	numOutputs := len(values)
+	if keep > 0 {
+		numOutputs++
+	}
+	if btc.segwit {
+		splitTxSizeWithoutInputs += uint64(dexbtc.P2WPKHOutputSize * numOutputs)
+	} else {
+		splitTxSizeWithoutInputs += uint64(dexbtc.P2PKHOutputSize * numOutputs)
+	}
+	enough := func(inputSize, sum uint64) (bool, uint64) {
+		splitTxFee := (splitTxSizeWithoutInputs + inputSize) * splitTxFeeRate
+		req := totalOutputRequired + splitTxFee
+		return sum >= req, sum - req
+	}
+	fundSplitCoins, _, _, _, inputsSize, totalIn, err := btc.fundInternal(keep, 0, true, enough)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	if maxLock > 0 {
+		totalSize := inputsSize + splitTxSizeWithoutInputs
+		if totalOutputRequired+(totalSize*splitTxFeeRate) > maxLock {
+			// This will only happen if we are right at the edge. If maxLock > totalOutputRequired,
+			// we would not get this far.
+			return nil, nil, 0, fmt.Errorf("amount required for split > maxLock")
+		}
+	}
+
+	baseTx, _, _, err := btc.fundedTx(fundSplitCoins)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	outputAddresses := make([]btcutil.Address, len(values))
+	for i, req := range requiredForOrders {
+		changeAddr, err := btc.node.changeAddress()
+		if err != nil {
+			return nil, nil, 0, err
+		}
+		outputAddresses[i] = changeAddr
+		script, err := txscript.PayToAddrScript(changeAddr)
+		if err != nil {
+			return nil, nil, 0, err
+		}
+		baseTx.AddTxOut(wire.NewTxOut(int64(req), script))
+	}
+
+	changeAddr, err := btc.node.changeAddress()
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	tx, err := btc.sendWithReturn(baseTx, changeAddr, totalIn, totalOutputRequired, splitTxFeeRate)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	txHash := tx.TxHash()
+
+	coins := make([]asset.Coins, len(values))
+	redeemScripts := make([][]dex.Bytes, len(values))
+	ops := make([]*output, len(values))
+	for i := range coins {
+		coins[i] = asset.Coins{newOutput(&txHash, uint32(i), uint64(tx.TxOut[i].Value))}
+		redeemScripts[i] = []dex.Bytes{tx.TxOut[i].PkScript}
```

Comment:
A redeem script is the script for a p2sh address. p2pkh addresses have no redeem script.